Tuturial read me (#1)
viktor-platform/sample-dynamo: README.md

# Tuturial dynamo 

> Welcome to this tutorial on integrating Dynamo with VIKTOR! VIKTOR is a platform dedicated to the Engineering and Construction industry, providing a seamless integration with 3rd party software. In this tutorial, we will explore how to render a basic house with Dynamo and visualize it in VIKTOR. To get started, the user will provide parameters such as the number of houses, width, depth, number of floors, and heights for floors and roofs. The Viktor worker will then compute the Dynamo model using the command-line interface included within Dynamo Sandbox. The geometry of the model is generated using either Autodesk Revit or FormIt. The geometry JSON is then converted to a mesh, which is rendered and visualized in Viktor. In this tutorial, we will cover the step-by-step process of setting up a VIKTOR app with Dynamo integration we will cover:

Author: ```suggestion
Welcome to this tutorial on integrating Dynamo with VIKTOR! VIKTOR is a platform dedicated to the Engineering and Construction industry, providing a seamless integration with 3rd party software. In this tutorial, we will explore how to render a basic house with Dynamo and visualize it in VIKTOR. To get started, the user will provide parameters such as the number of houses, width, depth, number of floors, and heights for floors and roofs. The Viktor worker will then compute the Dynamo model using the command-line interface included within Dynamo Sandbox. The geometry of the model is generated using either Autodesk Revit or FormIt. The geometry JSON is then converted to a mesh, which is rendered and visualized in VIKTOR. In this tutorial, we will cover the step-by-step process of setting up a VIKTOR app with Dynamo integration we will cover:
```

Step 1. [Set up a VIKTOR app with Dynamo integration](#1-creating-a-simple-template)

Step 2. [Setting up worker](#2-fill-the-template-with-data)

> By the end of this tutorial, you will have created a simple viktor application that creates a geometry and data view of a simple house, see gif below:

Author: ```suggestion
By the end of this tutorial, you will have created a simple VIKTOR application that creates a geometry and data view of a simple house, see GIF below:
```

![My Image](Read_me_files/Animation.gif)

## Pre-requisites

:::note Prerequisites
- You completed [Create your first app](/docs/getting-started/create-first-app) section.
- You have some experience with reading Python code
:::

During the tutorial, we added some links to additional information; but don't let them distract you too much. Stay

> focused on completing the tutorial. After this, you will know everything you need to create an app which includes Dynamo integration.

Author: ```suggestion
focused on completing the tutorial. After this, you will know everything you need to create an app which includes integrating with Dynamo.
```

## Step 1: Set up a VIKTOR app with Dynamo integration

## Step 1.1 Set up the VIKTOR app structure 
Let us assume that you have not yet set up a VIKTOR app yet. To set up a new project, use the following command in 
the terminal:

```

> viktor-cli create-app my-Dynamo-app

Author: ```suggestion
viktor-cli create-app my-dynamo-app
```

My preference goes for lowercase project names

```

This will set up a project/directory with the starting code needed to create a VIKTOR app.

In the `app.py` file, you will see the following code:

```python
from viktor import ViktorController
from viktor.parametrization import ViktorParametrization
class Parametrization(ViktorParametrization):
class Controller(ViktorController):
    label = 'My Entity Type'
    parametrization = Parametrization
```

As explained in VIKTOR documentation, the `Parametrization` class is used to define all input parameters you will 
find on the left-hand side of the VIKTOR application editor, and the `Controller` class is used define all logic of 
the application editor.

### Step 1.2 Define the input fields

> The dynamo file has got the following input paramaters.

Author: ```suggestion
The Dynamo file has got the following input paramaters.
```

| Number of houses | {{n_houses}}  |
| Width  | {{width}}  |
| Depth | {{depth}}  |
| Number of floors  | {{n_floors}}  |
| Height floor  | {{height_floor}}  |
| Height roof  | {{height_roof}}  |

We will use the numberfields to dynamically change those parameters. This is how it 
will look like in the `Parametrization` class:

```python
...
from viktor.parametrization import ViktorParametrization, NumberField
class Parametrization(ViktorParametrization):
    # Input fields
    number_of_houses = NumberField("Number of houses", max=8.0, min=1.0, variant='slider', step=1.0, default=3.0)
    number_of_floors = NumberField("Number of floors", max=5.0, min=1.0, variant='slider', step=1.0, default=2.0)
    depth = NumberField("Depth [m]", max=10.0, min=5.0, variant='slider', step=1.0, default=8.0)
    width = NumberField("Width [m]", max=6.0, min=4.0, variant='slider', step=1.0, default=5.0)
    height_floor = NumberField("Height floor", max=3.0, min=2.0, variant='slider', step=0.1, default=2.5, suffix='m')
    height_roof = NumberField("Height roof", max=3.0, min=2.0, variant='slider', step=0.1, default=2.5, suffix='m')
```

### Step 1.3 Define the app logic to update the dynamo file

>  This step, we will define the code to take the parameters as input, update the dynamo file, and return the updated files. This will be done by creating a `staticmethod`  in the `controller` class, similar to what is described in the [Generic integration](https://docs.viktor.ai/docs/create-apps/software-integrations/dynamo) section of the VIKTOR documentation.

Author: ```suggestion
 This step, we will define the code to take the parameters as input, update the Dynamo file, and return the updated files. This will be done by creating a `staticmethod`  in the `controller` class, similar to what is described in the [Generic integration](https://docs.viktor.ai/docs/create-apps/software-integrations/dynamo) section of the VIKTOR documentation.
```

> The following code will update the nodes of the dynamo file and generate an input file  with the parameters from the parametrization class:

Author: ```suggestion
The following code will update the nodes of the Dynamo file and generate an input file  with the parameter values as defined in the Parametrization class:
```

```python
from pathlib import Path
from viktor import File
from viktor.external.dynamo import DynamoFile

@staticmethod
def update_model(params) -> Tuple[File, DynamoFile]:
    """This method updates the nodes of the dynamo file with the parameters
    from the parametrization class."""

    # First the path to the dynamo file is specified and loaded
    file_path = Path(__file__).parent / "dynamo_model_sample_app.dyn"
    _file = File.from_path(file_path)
    dyn_file = DynamoFile(_file)

    # Update dynamo file with parameters from user input
    dyn_file.update("Number of houses", params.number_of_houses)
    dyn_file.update("Number of floors", params.number_of_floors)
    dyn_file.update("Depth", params.depth)
    dyn_file.update("Width", params.width)
    dyn_file.update("Height floor", params.height_floor)
    dyn_file.update("Height roof", params.height_roof)

    # generate updated file
    input_file = dyn_file.generate()

    return input_file, dyn_file
```
Let us go through the above mentioned logic:

> 1. Retrieve the input files for the analysis, in this case the `dynamo_model_sample_app.dyn` file, and create a DynamoFile object instantiated from the `dynamo_model_sample_app.dyn` file.

Author: ```suggestion
1. Retrieve the input files for the analysis, in this case the `dynamo_model_sample_app.dyn` file, and create a `DynamoFile` object instantiated from the `dynamo_model_sample_app.dyn` file.
```

> 2. With the update method, the value of input nodes can be updated.

Author: ```suggestion
2. With the `update` method, the value of input nodes can be updated.
```

3. When all inputs have been updated as desired, the generate method can be used to generate an updated `File` object.

> ### Step 1.4 Define the app logic to convert the output.xml to a data_group

Author: ```suggestion
### Step 1.4 Define the app logic to convert the output.xml to a DataGroup
```

> In this step, we will define code to extract data from the dynamo file and output file. The dynamo file is used to get the node id, and the output.xml file is used to get the values. We will be creating another `staticmethod`  in the `controller`    class. See code below:

Author: ```suggestion
In this step, we will define code to extract data from the Dynamo file and output file. The Dynamo file is used to get the node id, and the output.xml file is used to get the values. We will be creating another `staticmethod`  in the `Controller`    class. See code below:
```

    output_id_total_cost = input_file.get_node_id("(OUTPUT) Total cost")
    output_id_mki = input_file.get_node_id("(OUTPUT) MKI")
    output_id_co2 = input_file.get_node_id("(OUTPUT) CO2")

    # Collect the numerical results from the output file using the collected ids
    with output_file.open_binary() as f:
        floor_area = get_dynamo_result(f, id_=output_id_floor_area)
        total_cost = get_dynamo_result(f, id_=output_id_total_cost)
        mki = get_dynamo_result(f, id_=output_id_mki)
        co2 = get_dynamo_result(f, id_=output_id_co2)

    # Add values to a structured data group
    data_group = DataGroup(
        DataItem(label="Floor area", value=round(float(floor_area), 2), suffix="m²"),
        DataItem(label="Total cost", value=round(float(total_cost), 2), suffix="€"),
        DataItem(label="MKI", value=round(float(mki), 2)),
        DataItem(label="CO₂ emission", value=round(float(co2), 2), suffix="ton CO₂"),
    )
    return data_group
```
Let us go through the above mentioned logic:
1. Get the node_id, which corresponds to the same node id as the input file.
2. Collect the numerical results.

> 3. We create a structured data group from the collected numerical results using the `DataGroup` and `DataItem` classes.

Author: I'm not sure if the reader will understand what we mean by a structured DataGroup. Maybe good to mention that for more information on this, refer to:

https://docs.viktor.ai/docs/create-apps/results-and-visualizations/data-and-tables

### Step 1.5 generate a geometry and data view

> We create a new `GeometryAndDataView` method insdie the `controller` class. This method will enable us to show the 3D model of the house along with its data. Here is the code:

Author: ```suggestion
We create a new `GeometryAndDataView` method insdie the `Controller` class. This method will enable us to show the 3D model of the house along with its data. Here is the code:
```

from viktor.views import GeometryView, GeometryResult

@GeometryAndDataView("Building 3D", duration_guess=5)
def geometry_and_data_view(self, params, **kwargs):
    """The endpoint that initiates the logic to visualize the geometry and data executed
    and retrieved from a Dynamo script."""
    # Step 1: Update model
    input_file, dynamo_file = self.update_model(params)

    # Step 2: Running analyses
    files = [
        ('input.dyn', input_file),
    ]

    generic_analysis = GenericAnalysis(files=files, executable_key="dynamo",
                                        output_filenames=["output.xml", "geometry.json"])
    generic_analysis.execute(timeout=60)

    # Step 3: Processing geometry
    geometry_file = generic_analysis.get_output_file('geometry.json', as_file=True)
    glb_file = convert_geometry_to_glb(geometry_file)

    # Step 4: Process numerical output
    output_file = generic_analysis.get_output_file('output.xml', as_file=True)
    data_group = self.convert_dynamo_file_to_data_items(dynamo_file, output_file)

    return GeometryAndDataResult(geometry=glb_file, data=data_group)
```
Let us go through the above mentioned logic:

> 1. Update the dynamo file with the `update_model` method(see step 1.3)

Author: ```suggestion
1. Update the Dynamo file with the `update_model` method(see step 1.3)
```
